### PR TITLE
feat: [PLG-2407]: Added modules to subscriptionDetailDTO

### PR DESCRIPTION
**Pull request by thomas-grasman**
## Describe your changes
Added a collection of module types to subscription detail DTO and a helper method to extract the module types from the subscription line items' metadata.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48265)

### CODE REVIEW

**Comment by thomas-grasman** [comment]
trigger uts

File: 935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java [hunk 1]
```diff
@@ -423,6 +426,21 @@ public InvoiceDetailDTO finalizeInvoice(String invoiceId) {
...
 
+  private String getModuleType(SubscriptionItem subscriptionItem) {
+    return subscriptionItem.getPrice().getMetadata().get(STRIPE_MODULE_TYPE_KEY);
+  }
+
+  private Set<ModuleType> getModuleTypes(Subscription subscription) {
+    Set<ModuleType> moduleTypes = new HashSet<>();
+    subscription.getItems().getData().stream().forEach((SubscriptionItem subscriptionItem) -> {
+      String moduleType = getModuleType(subscriptionItem);
+      if (moduleType != null && !moduleTypes.contains(moduleType)) {
```
Comment (Xin-Harness) [inline 935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java:437]: There may have some issue, the moduleType from line 436 is a String type but the moduleTypes set contains the enum type ModuleType, so the moduleTypes.contains(moduleType) doesn't look so right, not sure why the IDE didn't complain about it. Suggest to convert to Enum then do the condition check.

File: 950-subscription-beans/src/main/java/io/harness/subscription/dto/SubscriptionDetailDTO.java [hunk 2]
```diff
@@ -30,4 +33,5 @@ public class SubscriptionDetailDTO {
   private String latestInvoice;
   private InvoiceDetailDTO latestInvoiceDetail;
   @Optional private List<ItemDTO> items;
+  @Optional private Set<ModuleType> modules;
```
Comment (Xin-Harness) [inline 950-subscription-beans/src/main/java/io/harness/subscription/dto/SubscriptionDetailDTO.java:36]: Shall we keep the naming convention to moduleTypes as more precise and unify?

**Review by Xin-Harness — CHANGES_REQUESTED** [review]

**Comment by thomas-grasman** [comment]
trigger uts

**Comment by thomas-grasman** [comment]
trigger uts

**Review by thomas-grasman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48265)* status: 2 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java` line 437 at r1](https://reviewable.io/reviews/harness/harness-core/48265#gh-1204777119:-NWIJxgJ-LqY3g0ONywb:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/13fe997434ba4edf4af30b27465207012235eca1/935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java#L437)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

There may have some issue, the moduleType from line 436 is a String type but the moduleTypes set contains the enum type ModuleType, so the moduleTypes.contains(moduleType) doesn't look so right, not sure why the IDE didn't complain about it. Suggest to convert to Enum then do the condition check.  
</blockquote></details>

Done.
___
*[`950-subscription-beans/src/main/java/io/harness/subscription/dto/SubscriptionDetailDTO.java` line 36 at r1](https://reviewable.io/reviews/harness/harness-core/48265#gh-1204779141:-NWIJwLPAD7b7Z7WfcRj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/13fe997434ba4edf4af30b27465207012235eca1/950-subscription-beans/src/main/java/io/harness/subscription/dto/SubscriptionDetailDTO.java#L36)):*
<details><summary><i>Previously, Xin-Harness (Xin Shao) wrote…</i></summary><blockquote>

Shall we keep the naming convention to moduleTypes as more precise and unify? 
</blockquote></details>

Done.

**Comment by thomas-grasman** [comment]
trigger uts

**Review by Xin-Harness — APPROVED** [review]

**Review by kapil-harness — APPROVED** [review]
@thomas-grasman you can add a UT for this logic in next PR, thanks!

> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48265#-:-NWJHWV78a4YeLcos0sD:b4hed50)* status: 2 file(s) to review. 2. unresolved discussions.

**Pull request closed without merging**